### PR TITLE
Disable tensor.from_elements from being a scalar outline

### LINKED ISSUE #14255: Control flow / RNG operations fail to compile due to `tensor.from_elements` operation
We are failing to compile some models with control flow due to the `tensor.from_elements` operator. See the IR below.

Overall it results in a `tensor.from_elements` operation existing and being unable to compile away during compilation
 to ConvertToLLVM. It is likely this should be cleaned up earlier in the pipeline through detensorization.

Input IR:
```
  func.func public @main(%arg0: tensor<i32> {mhlo.sharding = "{replicated}"}, %arg1: tensor<2x2xi32>, %arg2: tensor<2x2xi32>) -> (tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>) {
    %0:3 = stablehlo.while(%iterArg = %arg0, %iterArg_0 = %arg1, %iterArg_1 = %arg2) : tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>
     cond {
      %1 = stablehlo.constant dense<5> : tensor<i32>
      %2 = stablehlo.compare  LT, %iterArg, %1,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
      stablehlo.return %2 : tensor<i1>
    } do {
      %1 = stablehlo.constant dense<1> : tensor<i32>
      %2 = stablehlo.add %iterArg, %1 : tensor<i32>
      stablehlo.return %2, %iterArg_1, %iterArg_0 : tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>
    }
    return %0#0, %0#1, %0#2 : tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>
  }
```

### PR DESCRIPTION
**Pull request by rsuderman**
`tensor.from_elements` should remain as a `flow` operation and should not be outlined when forming scalar dispatches.

### CODE REVIEW

**Comment by rsuderman** [comment]
@phoenix-meadowlark this should fix the `tensor.from_elements` issue

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I am not sure this is a good idea. All the elements used to create the tensor will get moved to the host to create the tensor. That seems against what this pass is doing

**Comment by rsuderman** [comment]
@MaheshRavishankar 
It may be that we are missing a rewrite pattern within our scalar dispatches then. See the following program that fails to compile:

```
  func.func public @main(%arg0: tensor<i32> {mhlo.sharding = "{replicated}"}, %arg1: tensor<2x2xi32>, %arg2: tensor<2x2xi32>) -> (tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>) {
    %0:3 = stablehlo.while(%iterArg = %arg0, %iterArg_0 = %arg1, %iterArg_1 = %arg2) : tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>
     cond {
      %1 = stablehlo.constant dense<5> : tensor<i32>
      %2 = stablehlo.compare  LT, %iterArg, %1,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
      stablehlo.return %2 : tensor<i1>
    } do {
      %1 = stablehlo.constant dense<1> : tensor<i32>
      %2 = stablehlo.add %iterArg, %1 : tensor<i32>
      stablehlo.return %2, %iterArg_1, %iterArg_0 : tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>
    }
    return %0#0, %0#1, %0#2 : tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>
  }
  ```
  
  Results in:
  ```
  /tmp/test.mlir:9:12: error: 'tensor.from_elements' op unhandled tensor operation
      %2 = stablehlo.add %iterArg, %1 : tensor<i32>
           ^
/tmp/test.mlir:1:3: note: called from
  func.func public @main(%arg0: tensor<i32> {mhlo.sharding = "{replicated}"}, %arg1: tensor<2x2xi32>, %arg2: tensor<2x2xi32>) -> (tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>) {
  ^
/tmp/test.mlir:9:12: note: see current operation: %5 = "tensor.from_elements"(%4) : (i32) -> tensor<i32>
      %2 = stablehlo.add %iterArg, %1 : tensor<i32>
           ^
/tmp/test.mlir:9:12: error: 'func.func' op failed to create tensor equivalance classes
      %2 = stablehlo.add %iterArg, %1 : tensor<i32>
           ^
/tmp/test.mlir:1:3: note: called from
  func.func public @main(%arg0: tensor<i32> {mhlo.sharding = "{replicated}"}, %arg1: tensor<2x2xi32>, %arg2: tensor<2x2xi32>) -> (tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>) {
  ```

**Comment by benvanik** [comment]
agreed with mahesh we don't want these to stay outside of dispatches

**Comment by MaheshRavishankar** [comment]
Its hard to follow from just the error. Could you post the IR before the `FormScalarDispatches` pass?

**Comment by rsuderman** [comment]
So the problem does not occur in OutlineScalarDispatches - the issue is the only rewriter we have for `tensor.from_elements` is the conversion to `flow.splat`. My understanding is that we cannot perform that within a dispatch region. Somewhere within our dispatch region legalization we need to specify how we handle `tensor.from_elements` to scalar elements.

**Comment by MaheshRavishankar** [comment]
> So the problem does not occur in OutlineScalarDispatches - the issue is the only rewriter we have for `tensor.from_elements` is the conversion to `flow.splat`. My understanding is that we cannot perform that within a dispatch region. Somewhere within our dispatch region legalization we need to specify how we handle `tensor.from_elements` to scalar elements.

Still the IR before this pass would be useful. Basically need to understand what the IR looks like before this pass. That pass should evolve to also move loop computations  into dispatches. This might be the issue here.

**Comment by rsuderman** [comment]
Here is the IR before the outline scalar dispatch regions:
```
func.func @main(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view) -> (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view) attributes {iree.abi.stub} {
  %c1_i32 = arith.constant 1 : i32
  %c5_i32 = arith.constant 5 : i32
  %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<i32>
  %1 = hal.tensor.import %arg1 "input 1" : !hal.buffer_view -> tensor<2x2xi32>
  %2 = hal.tensor.import %arg2 "input 2" : !hal.buffer_view -> tensor<2x2xi32>
  %extracted = tensor.extract %0[] : tensor<i32>
  cf.br ^bb1(%extracted, %0, %1, %2 : i32, tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>)
^bb1(%3: i32, %4: tensor<i32>, %5: tensor<2x2xi32>, %6: tensor<2x2xi32>):  // 2 preds: ^bb0, ^bb2
  %7 = arith.cmpi slt, %3, %c5_i32 : i32
  cf.cond_br %7, ^bb2, ^bb3
^bb2:  // pred: ^bb1
  %8 = arith.addi %3, %c1_i32 : i32
  %from_elements = tensor.from_elements %8 : tensor<i32>
  cf.br ^bb1(%8, %from_elements, %6, %5 : i32, tensor<i32>, tensor<2x2xi32>, tensor<2x2xi32>)
^bb3:  // pred: ^bb1
  %9 = hal.tensor.export %4 "output 0" : tensor<i32> -> !hal.buffer_view
  %10 = hal.tensor.export %5 "output 1" : tensor<2x2xi32> -> !hal.buffer_view
  %11 = hal.tensor.export %6 "output 2" : tensor<2x2xi32> -> !hal.buffer_view
  return %9, %10, %11 : !hal.buffer_view, !hal.buffer_view, !hal.buffer_view
}
```
